substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/SubstrateJVM.java

```diff
@@ -651,6 +681,30 @@ public Object getConfiguration(Class<? extends Event> eventClass) {
         return DynamicHub.fromClass(eventClass).getJfrEventConfiguration();
     }
 
+    public void exclude(Thread thread) {
+        JfrThreadLocal jfrThreadLocal = (JfrThreadLocal) getThreadLocal();
+        jfrThreadLocal.exclude(thread);
+    }
+
+    public void include(Thread thread) {
+        JfrThreadLocal jfrThreadLocal = (JfrThreadLocal) getThreadLocal();
+        jfrThreadLocal.include(thread);
+    }
+
+    public boolean isExcluded(Thread thread) {
+        // in jdk 17 and jdk19 the argument is only ever the current thread.
+        if (!thread.equals(Thread.currentThread())) {
```

Comment:
But would it be safe to access that other thread's `javaEventWriter` thread local later in the method?

Reply:
`javaEventWriter` is a `FastThreadLocal`, so it is unsafe to access from other threads (it needs the `IsolateThread` and this data structure may be freed at any time if the owning thread exits).
I don't fully understand how the `excluded` flag in `jdk.jfr.internal.event.EventWriter` needs to interact with the virtual threads. In HotSpot, there are 3 different excluded flags (`EventWriter.excluded`, `_vthread_excluded`, and `_jvm_thread_excluded`), so we might miss something here.

Reply:
I looks like in the hotspot code, when a java thread is excluded (whether it isa platform thread or a vthread) it will set the `excluded` flag in the thread's corresponding `jdk.jfr.internal.event.EventWriter`.  When this flag is set, that EventWriter wont make any more java events.   Seems like`_vthread_excluded` is for vthreads while `_jvm_thread_excluded` is for Java platform threads and native threads too.

I'll add back the check that ensures a thread can only exclude itself. This way it is safe to access the `javaEventWriter` thread local.